ledger/src/blockstore_meta.rs

```diff
@@ -269,6 +303,278 @@ impl ShredIndex {
+    {
+        use serde::ser::SerializeTuple;
+
+        let mut tuple = serializer.serialize_tuple(2)?;
+        // SAFETY: This is safe because:
+        // 1. Memory initialization & layout:
+        //    - index is a fixed-size array [u64; MAX_U64S_PER_SLOT] fully initialized
+        //      at construction and never contains uninitialized memory
+        //    - array elements are contiguous with no padding
+        //
+        // 2. Size & alignment:
+        //    - size_of::<[u64; MAX_U64S_PER_SLOT]> is exactly (8 * MAX_U64S_PER_SLOT) bytes
+        //    - &self.index is u64-aligned, which satisfies u8 alignment
+        //
+        // 3. Lifetime:
+        //    - slice lifetime is tied to &self and is read-only
+        //
+        // Note: Deserialization will validate the byte length and safely reconstruct the u64 array
+        tuple.serialize_element(&serde_bytes::Bytes::new(unsafe {
+            std::slice::from_raw_parts(
+                &self.index as *const _ as *const u8,
+                std::mem::size_of::<ShredIndexV2Inner>(),
+            )
+        }))?;
+        tuple.serialize_element(&self.num_shreds)?;
+        tuple.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for ShredIndexV2 {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let (bytes, num_shreds) = <(&[u8], usize)>::deserialize(deserializer)?;
+        // Accept input smaller than our current fixed array size to maintain compatibility with older
+        // versions that had fewer shreds per slot. This is safe because smaller sets are stored in
+        // our fixed array with any remaining space automatically zeroed. This approach ensures we can
+        // read data from any previous version while enforcing a maximum size limit.
+        if bytes.len() > std::mem::size_of::<ShredIndexV2Inner>() {
+            return Err(serde::de::Error::custom("input too large"));
+        }
+        let mut index = [0u64; NUM_U64_PER_SHRED_INDEX_V2];
+        bytes.chunks_exact(8).enumerate().for_each(|(i, chunk)| {
+            // Unwrap is safe because `chunks_exact` guarantees the length
+            index[i] = u64::from_ne_bytes(chunk.try_into().unwrap());
+        });
+        Ok(Self { index, num_shreds })
+    }
+}
+
+impl ShredIndexV2 {
+    pub fn num_shreds(&self) -> usize {
+        self.num_shreds
+    }
+
+    fn index_and_mask(index: u64) -> (usize, u64) {
+        // index / 64
+        let word_idx = (index >> 6) as usize;
+        // index % 64
+        let bit_idx = index & 63;
+        let mask = 1 << bit_idx;
+        (word_idx, mask)
+    }
+
+    #[allow(unused)]
+    fn remove(&mut self, index: u64) {
+        assert!(
+            index < MAX_DATA_SHREDS_PER_SLOT as u64,
+            "index out of bounds. {index} >= {MAX_DATA_SHREDS_PER_SLOT}"
+        );
+
+        let (word_idx, mask) = Self::index_and_mask(index);
+
+        if self.index[word_idx] & mask != 0 {
```

Comment:
what guarantees that `word_idx` is less than `self.index.len()` here?
this will panic if that presumption fails.